input/posts/2020-07-08-grpc-reflection-in-net.md

```diff
@@ -0,0 +1,226 @@
+Title: gRPC Server Reflection in the .NET world
+Published: 2020-07-08
+Tags: 
+- .NET
+- gRPC
+- Protocol Buffers
+---
+> gRPC Server Reflection is an optional extension for servers to assist clients in runtime construction of requests without having stub information precompiled into the client.
+> 
+> -- <cite>[gRPC Server Reflection Protocol](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md)</cite>
+
+The primary use case for gRPC Server Reflection is debugging and testing tools. Think cURL or Postman, but for gRPC. There's however another interesting use case, especially in the .NET world where we have code-first support through [protobuf-net.Grpc](https://github.com/protobuf-net/protobuf-net.Grpc).
+
+If you have been using WCF, you are probably familiar with code-first approach, and exposing metadata about the service using MEX or WSDL endpoints. These metadata endpoints could then be used to generate client stubs. Additionally, if you created a standard SOAP service, to be consumed over HTTP, interoperability with other programming languages was also possible by consuming the WSDL document. This is similar to what many of us do today with ASP.NET Core and OpenAPI. We create the controllers and models in C#, generate OpenAPI definition using [Swashbuckle](https://github.com/domaindrivendev/Swashbuckle.AspNetCore) and then generate clients in TypeScript using [Autorest](https://github.com/Azure/autorest).
+
+While Google's way of working with protobuf and gRPC have always been contract-first, [Marc Gravell's](https://github.com/mgravell) [protobuf-net](https://github.com/protobuf-net/protobuf-net) have always had code-first development as first priority. With Google's protobuf, you define your services and messages in .proto format and then use the protobuf compiler (protoc) to generate messages, client- and server stubs in the language of your choice. With protobuf-net, however, you can start out with C#, F# or VB.NET code. There's no need for .proto file. Code-first with protobuf-net is great if you don't need interoperability with other languages outside the .NET ecosystem, and even if you did, there have been ways to to get the contract generated in .proto format for message types. However, up until recently, if you needed the proto contract for gRPC services you were on your own. This was however fixed last weekend.
```

Comment:
I really wish we could indicate exact words for these comments...oh well - a couple more suggestions: "to generate messages, client**,** and server stubs"..."**However, with** protobuf-net you can"..."There's no need for **a** .proto file"